File: assets/code.java

import java.awt.Color;
import java.awt.image.BufferedImage;
import java.io.File;
import java.io.IOException;
import java.util.Random;
import javax.imageio.ImageIO;

public class ImageProcessing {

  private static final Random RANDOM = new Random();

  public static void main(String[] args) throws IOException {
    // Read the input image
    BufferedImage inputImage = ImageIO.read(new File("assets/input.jpg"));

    // Apply salt and pepper noise to the image
    applySaltAndPepperNoise(inputImage, 0.05, 0.05);

    // Apply a median filter to the image
    applyMedianFilter(inputImage, 3);

    // Apply Prewitt Operation
    applyPrewittOperation(inputImage);

    // Write the output image to a file
    ImageIO.write(inputImage, "jpg", new File("assets/output.jpg"));
  }

  private static void applySaltAndPepperNoise(BufferedImage image, double whiteProbability, double blackProbability) {
    for (int x = 0; x < image.getWidth(); x++) {
      for (int y = 0; y < image.getHeight(); y++) {
        double randomNumber = RANDOM.nextDouble();
        if (randomNumber < whiteProbability) {
          // Set the pixel to white
          image.setRGB(x, y, Color.WHITE.getRGB());
        } else if (randomNumber < whiteProbability + blackProbability) {
          // Set the pixel to black
          image.setRGB(x, y, Color.BLACK.getRGB());
        }
      }
    }
  }

  private static void applyMedianFilter(BufferedImage image, int kernelSize) {
    int[][][] pixels = new int[image.getWidth() + kernelSize * 2][image.getHeight() + kernelSize * 2][3];
    for (int x = 0; x < image.getWidth(); x++) {
      for (int y = 0; y < image.getHeight(); y++) {
        for (int rgbChannel = 0; rgbChannel < 3; rgbChannel++) {
          pixels[x + kernelSize][y + kernelSize][rgbChannel] = image.getRGB(x, y) >> (8 * rgbChannel) & 0xff;
        }
      }
    }

    for (int x = 0; x < image.getWidth(); x++) {
      for (int y = 0; y < image.getHeight(); y++) {
        for (int rgbChannel = 0; rgbChannel < 3; rgbChannel++) {
          int[] sortedValues = new int[kernelSize * kernelSize];
          int index = 0;
          for (int i = -kernelSize / 2; i <= kernelSize / 2; i++) {
            for (int j = -kernelSize / 2; j <= kernelSize / 2; j++) {
              sortedValues[index++] = pixels[x + kernelSize + i][y + kernelSize + j][rgbChannel];
            }
          }
          Arrays.sort(sortedValues);
          image.setRGB(x, y, new Color(sortedValues[kernelSize * kernelSize / 2], sortedValues[kernelSize * kernelSize / 2], sortedValues[kernelSize * kernelSize / 2]).getRGB());
        }
      }
    }
  }

  private static void applyPrewittOperation(BufferedImage image) {
    int[][][] kernelX = { { -1, 0, 1 }, { -1, 0, 1 }, { -1, 0, 1 } };
    int[][][] kernelY = { { -1, -1, -1 }, { 0, 0, 0 }, { 1, 1, 1 } };

    int[][][] gradientX = new int[image.getWidth()][image.getHeight()][3];
    int[][][] gradientY = new int[image.getWidth()][image.getHeight()][3];

    for (int x = 0; x < image.getWidth(); x++) {
      for (int y = 0; y < image.getHeight(); y++) {
        for (int rgbChannel = 0; rgbChannel < 3; rgbChannel++) {
          gradientX[x][y][rgbChannel] = convolve(image, kernelX, x, y, rgbChannel);
          gradientY[x][y][rgbChannel] = convolve(image, kernelY, x, y, rgbChannel);
        }
      }
    }

    for (int x = 0; x < image.getWidth(); x++) {
      for (int y = 0; y < image.getHeight(); y++) {
        for (int rgbChannel = 0; rgbChannel < 3; rgbChannel++) {
          // Calculate the magnitude of the gradient
          int magnitude = (int) Math.sqrt(gradientX[x][y][rgbChannel] * gradientX[x][y][rgbChannel] + gradientY[x][y][rgbChannel] * gradientY[x][y][rgbChannel]);
          // Normalize the magnitude to the range [0, 255]
          magnitude = Math.min(magnitude, 255);
          // Set the pixel to the magnitude value
          image.setRGB(x, y, new Color(magnitude, magnitude, magnitude).getRGB());
        }
      }
    }
  }

  private static int convolve(BufferedImage image, int[][][] kernel, int x, int y, int rgbChannel) {
    int sum = 0;
    for (int i = -kernel.length / 2; i <= kernel.length / 2; i++) {
      for (int j = -kernel[0].length / 2; j <= kernel[0].length / 2; j++) {
        int pixelX = x + i;
        int pixelY = y + j;
        if (pixelX >= 0 && pixelX < image.getWidth() && pixelY >= 0 && pixelY < image.getHeight()) {
          sum += image.getRGB(pixelX, pixelY) >> (8 * rgbChannel) & 0xff * kernel[i + kernel.length / 2][j + kernel[0].length / 2][rgbChannel];
        }
      }
    }
    return sum;
  }
}

File: assets/code.py

import numpy as np
import cv2

class ImageProcessing:
    def __init__(self, path):
        self.image = cv2.imread(path)

    def apply_salt_and_pepper_noise(self, white_probability=0.05, black_probability=0.05):
        """Apply salt and pepper noise to the image.

        Args:
            white_probability: 0.05.
            black_probability: 0.05.
        """
        out = np.copy(self.image)

        # Generate a random array of 0s and 1s
        noise = np.random.choice([0, 1], size=self.image.shape, p=[white_probability, 1-white_probability])

        # Add noise to the image
        out[noise == 1] = 255
        out[noise == 0] = 0

        return out

    def apply_median_blur(self, kernel_size=3):
        """Apply a median filter to the image.

        Args:
            kernel_size: 3.
        """
        return cv2.medianBlur(self.image, kernel_size)

    def apply_prewitt_operation(self):
        """Apply Prewitt Operation with given path of image as 'assets' Use the OOP and Clean code principles write a ImageProcessing classand save the result image in the same path with the name of output.jpg."""

        # Convert the image to grayscale
        gray = cv2.cvtColor(self.image, cv2.COLOR_BGR2GRAY)

        # Apply the Prewitt operator to the image
        prewitt_x = cv2.Sobel(gray, cv2.CV_64F, 1, 0, ksize=3)
        prewitt_y = cv2.Sobel(gray, cv2.CV_64F, 0, 1, ksize=3)

        # Calculate the magnitude of the gradient
        magnitude = np.sqrt(prewitt_x**2 + prewitt_y**2)

        # Normalize the magnitude
        magnitude = magnitude / np.max(magnitude)

        # Save the output image
        cv2.imwrite('output.jpg', magnitude)

if __name__ == "__main__":
    image = ImageProcessing('path/to/image.jpg')

    # Apply salt and pepper noise to the image
    noisy_image = image.apply_salt_and_pepper_noise()

    # Apply a median filter to the image
    filtered_image = image.apply_median_blur()

    # Apply Prewitt Operation to the image
    image.apply_prewitt_operation()

    # Display the original, noisy, and filtered images
    cv2.imshow('Original Image', image.image)
    cv2.imshow('Noisy Image', noisy_image)
    cv2.imshow('Filtered Image', filtered_image)
    cv2.waitKey(0)
    cv2.destroyAllWindows()
